strawberryfields/engine.py

```diff
@@ -614,10 +615,19 @@ def run_async(self, program: Program, *, compile_options=None, **kwargs) -> Job:
         device = self.device_spec
 
         compiler_name = compile_options.get("compiler", device.default_compiler)
-        msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
-        self.log.info(msg)
 
-        program = program.compile(device=device, **compile_options)
+        if program.compiled_with is None:
+            msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
+            self.log.info(msg)
+
+            program = program.compile(device=device, **compile_options)
+
+        elif program.compiled_with != compiler_name:
+            msg = f"No compilation, program already compiled for {program.compiled_with}. Set recompile=true to compile with {compiler_name}."
+            self.log.info(msg)
```

Comment:
By storing both the device target _and_ the compiler for compiled programs, we can make the check a bit more fine grained here. For example,

```python
if program.compile_info is None:
    # program is uncompiled
    msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
    self.log.info(msg)
    program = program.compile(device=device, **compile_options)

elif program.compile_info[0] != device:
    # program was compiled for a different device
    if not recompile:
        raise ValueError(f"Cannot use program compiled for {program._compile_info[0].target} for target {self.target}")
    program = program.compile(device=device, **compile_options)

elif program.compile_info[1] != compiler_name:
    # program was compiled for the device, but using a different compiler
    if compile_options is None:
        # perhaps we put a optional validation check here? depends how much we trust the compilation
        program.compile(device=device, compiler="Xstrict")
    else:
        # user has explicitly provided compile_options to the engine, so best to apply these.
        program = program.compile(device=device, **compile_options)
```